cpp/bench/ann/src/raft/raft_cagra_hnswlib_wrapper.h

```diff
@@ -0,0 +1,216 @@
+  using type = int;
+};
+
+template <>
+struct hnsw_dist_t<int8_t> {
+  using type = int;
+};
+
+template <typename T, typename IdxT>
+class RaftCagraHnswlib : public ANN<T> {
+ public:
+  using typename ANN<T>::AnnSearchParam;
+
+  struct SearchParam : public AnnSearchParam {
+    int ef;
+    int num_threads = 1;
+  };
+
+  using BuildParam = raft::neighbors::cagra::index_params;
+
+  RaftCagraHnswlib(Metric metric, int dim, const BuildParam& param, int concurrent_searches = 1)
+    : ANN<T>(metric, dim), index_params_(param), dimension_(dim), handle_(cudaStreamPerThread)
+  {
+    index_params_.metric = parse_metric_type(metric);
+    RAFT_CUDA_TRY(cudaGetDevice(&device_));
+  }
+
+  ~RaftCagraHnswlib() noexcept {}
+
+  void build(const T* dataset, size_t nrow, cudaStream_t stream) final;
+
+  void set_search_param(const AnnSearchParam& param) override;
+
+  // TODO: if the number of results is less than k, the remaining elements of 'neighbors'
+  // will be filled with (size_t)-1
+  void search(const T* queries,
+              int batch_size,
+              int k,
+              size_t* neighbors,
+              float* distances,
+              cudaStream_t stream = 0) const override;
+
+  // to enable dataset access from GPU memory
+  AlgoProperty get_preference() const override
+  {
+    AlgoProperty property;
+    property.dataset_memory_type = MemoryType::HostMmap;
+    property.query_memory_type   = MemoryType::Host;
+    return property;
+  }
+  void save(const std::string& file) const override;
+  void load(const std::string&) override;
+
+ private:
+  void get_search_knn_results_(const T* query, int k, size_t* indices, float* distances) const;
+
+  raft::device_resources handle_;
+  BuildParam index_params_;
+  std::optional<raft::neighbors::cagra::index<T, IdxT>> index_;
+  int device_;
+  int dimension_;
+
+  std::unique_ptr<hnswlib::HierarchicalNSW<typename hnsw_dist_t<T>::type>> appr_alg_;
+  std::unique_ptr<hnswlib::SpaceInterface<typename hnsw_dist_t<T>::type>> space_;
+  int num_threads_;
+  std::unique_ptr<FixedThreadPool> thread_pool_;
+
+  Objective metric_objective_;
+};
+
+template <typename T, typename IdxT>
+void RaftCagraHnswlib<T, IdxT>::build(const T* dataset, size_t nrow, cudaStream_t)
+{
+  if (raft::get_device_for_address(dataset) == -1) {
+    auto dataset_view =
+      raft::make_host_matrix_view<const T, int64_t>(dataset, IdxT(nrow), dimension_);
+    index_.emplace(raft::neighbors::cagra::build(handle_, index_params_, dataset_view));
+    return;
+  } else {
+    auto dataset_view =
+      raft::make_device_matrix_view<const T, int64_t>(dataset, IdxT(nrow), dimension_);
+    index_.emplace(raft::neighbors::cagra::build(handle_, index_params_, dataset_view));
+    return;
+  }
+}
+
+template <typename T, typename IdxT>
+void RaftCagraHnswlib<T, IdxT>::set_search_param(const AnnSearchParam& param_)
+{
+  auto param        = dynamic_cast<const SearchParam&>(param_);
+  appr_alg_->ef_    = param.ef;
```

Comment:
Just have this call `Hnswlib<T,IdxT>::set_search_param(param_);`  (or pull the logic into a shared helper functino).